### ISSUE TITLE
Coding Error in _LoadTexture

### ISSUE THREAD

**Issue by suekieza** [Autodesk/maya-usd#89]
Hey there,

refactoring_sandbox branch, Pulled down today with USD 1d08054

This barrel works just fine in Solaris, and usdview. But imports black(no textures) in maya 2019.1
FYI: I've check and the textures are in the right place. Not sure why maya is looking for them here: /home/suesauer/Documents/maya-usd/lib/render/vp2RenderDelegate/material.cpp 

The error it gives in maya. 

Error CODE:
`// Breakdown for file: /home/suesauer/Documents/LOPS_DEMO_FILES/Library/Assets/Barrels/BarrelsAsset.usd
[-2.14748e+07%](0ms) LoadStage
  [-2.14748e+07%](0ms) PostLoadProcess
    [-2.14748e+07%](0ms) CreateTransformChains
    [-2.14748e+07%](0ms) HuntForNativePrims
    [-2.14748e+07%](0ms) PostImportLogic
    [-2.14748e+07%](0ms) CreatePrims
  [-2.14748e+07%](0ms) OpeningUsdStage
    [-2.14748e+07%](0ms) UsdStageOpen
    [-2.14748e+07%](0ms) OpenRootLayer
    [-2.14748e+07%](0ms) OpeningSessionLayer
 // 
setFilterScript "initialShadingGroup";
// Result: 0 // 
setFilterScript "initialParticleSE";
// Result: 0 // 
setFilterScript "defaultLightSet";
// Result: 1 // 
setFilterScript "defaultObjectSet";
// Result: 1 // 
ls -type mayaUsdProxyShapeBase -long;
// Result: |AL_usdmaya_Proxy|AL_usdmaya_ProxyShape // 
ls -type mayaUsdProxyShapeBase -long;
// Result: |AL_usdmaya_Proxy|AL_usdmaya_ProxyShape // 
// Error: Failed verification: ' image ' -- Coding Error in _LoadTexture at line 275 of /home/suesauer/Documents/maya-usd/lib/render/vp2RenderDelegate/material.cpp // 
// Error: Failed verification: ' image ' -- Coding Error in _LoadTexture at line 275 of /home/suesauer/Documents/maya-usd/lib/render/vp2RenderDelegate/material.cpp // 
// Error: Failed verification: ' image ' -- Coding Error in _LoadTexture at line 275 of /home/suesauer/Documents/maya-usd/lib/render/vp2RenderDelegate/material.cpp // 
timeField -edit -value `currentTime -query` TimeSlider|MainTimeSliderLayout|formLayout8|timeField1;
// Result: TimeSlider|MainTimeSliderLayout|formLayout8|timeField1 // 
timeField -edit -value `currentTime -query` TimeSlider|MainTimeSliderLayout|formLayout8|timeField1;
// Result: TimeSlider|MainTimeSliderLayout|formLayout8|timeField1 // `

![Screenshot from 2019-10-30 12-01-52](https://user-images.githubusercontent.com/53234220/67848297-09425700-fb0d-11e9-8356-d94a3a071efd.png)

**Comment by BigRoy** [comment]
It's failing [here](https://github.com/Autodesk/maya-usd/blob/bf777793512b1c45999aad1d5bf95f0b902e7ec7/lib/render/vp2RenderDelegate/material.cpp#L275) on verifying the loaded image using [`TF_VERIFY`](https://graphics.pixar.com/usd/docs/api/group__group__tf___diagnostic.html#gaf621af3398d45d5d1108db257abc0ebd).

There's a lack of debug information here to figure out why it is exactly failing, but it is clear that it cannot read the texture. Can you try and change the code to this and recompile:
```cpp
    GlfImageSharedPtr image = GlfImage::OpenForReading(path);
    if (!TF_VERIFY(image, "Unable to load texture: %s", path)) {
        return nullptr;
    }
```
Then it should at least log what path it is trying to load from, so we can detect whether it resolved the wrong path for the texture or whether it's just unable to load a texture path that is actually correct.

**Comment by suekieza** [comment]
OK so here is the "path" it finds:
`// Breakdown for file: /home/suesauer/Documents/LOPS_DEMO_FILES/Library/Assets/Barrels/BarrelsAsset.usd
[-2.14748e+07%](0ms) LoadStage
  [-2.14748e+07%](0ms) PostLoadProcess
    [-2.14748e+07%](0ms) CreateTransformChains
    [-2.14748e+07%](0ms) HuntForNativePrims
    [-2.14748e+07%](0ms) PostImportLogic
    [-2.14748e+07%](0ms) CreatePrims
  [-2.14748e+07%](0ms) OpeningUsdStage
    [-2.14748e+07%](0ms) UsdStageOpen
    [-2.14748e+07%](0ms) OpeningSessionLayer
    [-2.14748e+07%](0ms) OpenRootLayer

// Error: Failed verification: ' image ' -- Unable to load texture: 8	 -- Coding Error in _LoadTexture at line 275 of /home/suesauer/Documents/maya-usd/lib/render/vp2RenderDelegate/material.cpp // 
// Error: Failed verification: ' image ' -- Unable to load texture: ?	 -- Coding Error in _LoadTexture at line 275 of /home/suesauer/Documents/maya-usd/lib/render/vp2RenderDelegate/material.cpp // 
// Error: Failed verification: ' image ' -- Unable to load texture: ? -- Coding Error in _LoadTexture at line 275 of /home/suesauer/Documents/maya-usd/lib/render/vp2RenderDelegate/material.cpp // 
createModelPanelMenu modelPanel1;
createModelPanelMenu modelPanel2;
createModelPanelMenu modelPanel3;
createModelPanelMenu modelPanel4;
createOutlinerPanelMenu outlinerPanel1;
buildPanelPopupMenu scriptEditorPanel1;
updateWorkspacesList;
buildViewMenu($gMainWindowMenu);`

Side note this used to work in the main branch.

Thanks

**Comment by BigRoy** [comment]
Interesting. Are you able to share the barrel example file with its textures here?

**Comment by suekieza** [comment]
Sure thing, 

see link

**Comment by BigRoy** [comment]
Thanks. Seems I had to request access, so you should have an e-mail about that. 👍

**Comment by suekieza** [comment]
Excepted your request :D

**Comment by BigRoy** [comment]
Thanks. I can reproduce the issue here. I think the debugging code I wrote might also be wrong as it might fail on actually correctly writing out the `path` variable, which seems to be `std::string` and since the formatting of `TF_VERIFY` basically uses `printf` and thus it's likely [slightly more involved](https://stackoverflow.com/questions/36908994/using-sprintf-with-stdstring-in-c) or [here](https://stackoverflow.com/questions/5322216/how-to-print-a-string-in-c).

I'll investigate on my end a bit more.

**Comment by suekieza** [comment]
hmmmmm isnt UDIM a standard why to read uv tiles? maybe @HamedSabri-adsk can way in on this as well

**Comment by suekieza** [comment]
also it worked before with the master branch and standard AL it also works(displays textures just fine)

**Comment by BigRoy** [comment]
@suekieza I believe the master branch and standard AL rely on Hydra to display it. As such I quickly tested in the `Hydra (GL)` viewport renderer, and voila. The texture instantly shows up.

As such, the textures solely fail to load in the *Viewport 2.0 Delegate* as they get loaded in a different way.

**Comment by suekieza** [comment]
Excellent, so at least this is something that can be fixed/restored? Thanks for all the help by the way, much appreciated.

**Comment by suekieza** [comment]
Hey @HdC-adsk any news on this problem? Thanks Sue

**Comment by HdC-adsk** [comment]
Hey @suekieza No. I am working on other tasks currently and don't have time to look into it yet. I will let you know when I make progress on this task, hopefully later today. Thanks.

**Comment by HdC-adsk** [comment]
Hey @suekieza , I just requested the access of the file and will work on it once I get the permission. If the issue is blocking you, the easiest workaround is to temporarily remove the following line in mayaUSD.mod.

VP2_RENDER_DELEGATE_PROXY=1

I will let you know when it is fixed so you can restore this line.

**Comment by HdC-adsk** [comment]
@suekieza Just in case you didn't notice mail notification, could you please grant me with the access permission of your asset? I've sent request via danod_chen@hotmail.com.

**Comment by suekieza** [comment]
Hey there, could you send me your email again I have miss placed it some how

**Comment by BigRoy** [comment]
> Hey there, could you send me your email again I have miss placed it some how. Nevermind I see you are shared now

@suekieza if you're fine with sharing the files publicly you can also `.zip` them up and drag and drop them into a new comment in this issue. This would instantly allow anyone to debug and look into your issue if they have the time without the need to request access  to your Google drive link.

**Comment by HdC-adsk** [comment]
Hey @suekieza , Yes, I've downloaded the asset you shared, thanks. As much as we want to implement the support for UDIM textures, we also need to implement a list of VP2RenderDelegate features which have been already supported by Hydra Storm. Could you please help me to understand how frequent and important UDIM textures are in your workflow? This will help us to prioritize and plan the coming tasks accordingly. Thanks.

**Comment by HdC-adsk** [comment]
Tracked as MAYA-101853 internally.

**Comment by suekieza** [comment]
Hey @HdC-adsk so had some chats internally about this problem, The overall consensus is that we will indeed need UDIMs for our projects as it helps keep texture map sizes under control. We seem to use them on all our assets so workflow wise they are a core part of texturing for us. Thanks

**Comment by HdC-adsk** [comment]
Hi @suekieza . Thanks for the info. We've also arranged this issue for the current sprint.

**Comment by suekieza** [comment]
Hey there, Any updates on this?

**Comment by HdC-adsk** [comment]
Hi @suekieza ,

Sorry we haven't had progress on this yet. I understand it is a blocking issue to you, could you remove "VP2_RENDER_DELEGATE_PROXY" env var in mayaUSD.mod to workaround? This will enable the AL draw override solution which uses Hydra Storm backend, so UDIM texture should be back.

We definitely need to add UDIM texture support for VP2RenderDelegate, but at the same time we also have some more important issues that would impact more general workflows thus we need to fix those issues as early as possible. Hope it makes sense to you. Thanks.

**Comment by suekieza** [comment]
Sure does thanks for the response :D

**Comment by kxl-adsk** [comment]
@suekieza UDIM texture support was added as part of #538. It requires the latest supported commit SHA from USD dev branch. Can you please validate and close this issue?

**Comment by santosd** [comment]
@suekieza I tested your file, on the latest internal version and it seems to be working. Please give it a try and let us know your results.

**Comment by suesauer** [comment]
Cool thanks guy's will test tomorrow :D/

**Comment by HdC-adsk** [comment]
@suesauer Any update so we can move on to close it?

**Comment by suesauer** [comment]
As far as I can see its still not working....

`file -f -new;
// untitled // 
commandPort -securityWarning -name commandportDefault;
onSetCurrentLayout "Maya Classic";
// Loading Bifrost version 2.0.5.1
// Bifrost: Loading library: Amino, from: Autodesk.
// Bifrost: Loading library: AminoMayaTranslation, from: Autodesk.
// Bifrost: Loading library: bif, from: Autodesk.
// Bifrost: Loading library: bifrostObjectMayaTranslations, from: Autodesk.
// Bifrost: Loading library: geometries, from: Autodesk.
// Bifrost: Loading library: fluids, from: Autodesk.
// Bifrost: Loading library: particles, from: Autodesk.
// Bifrost: Loading library: file, from: Autodesk.
// Bifrost: Loading library: midgard, from: Autodesk.
// Bifrost: Loading library: nucleus, from: Autodesk.
// Bifrost: Loading library: simulation, from: Autodesk.
updateRenderOverride;
updateRenderOverride;
// Warning: file: /usr/autodesk/maya2020/scripts/startup/autoLoadPlugin.mel line 35: Loading plug-in "mtoh" has resulted in changes to the scene that may need to be saved. // 
// Error: XGen:  Failed to create directory: /usr/tmp/_____xgen/2020721/
( system = "No such file or directory")  // 
// Warning: file: /home/sue.sauer/maya/2020/prefs/filePathEditorRegistryPrefs.mel line 4: filePathEditor: Attribute 'aiImage.filename' is invalid or is not designated 'usedAsFilename'. // 
// Warning: file: /home/sue.sauer/maya/2020/prefs/filePathEditorRegistryPrefs.mel line 5: filePathEditor: Attribute 'aiPhotometricLight.aiFilename' is invalid or is not designated 'usedAsFilename'. // 
// Warning: file: /home/sue.sauer/maya/2020/prefs/filePathEditorRegistryPrefs.mel line 6: filePathEditor: Attribute 'aiStandIn.dso' is invalid or is not designated 'usedAsFilename'. // 
// Warning: file: /home/sue.sauer/maya/2020/prefs/filePathEditorRegistryPrefs.mel line 7: filePathEditor: Attribute 'aiVolume.filename' is invalid or is not designated 'usedAsFilename'. // 
// AbcExport v1.0 using Alembic 1.7.5 (built Feb  7 2018 15:59:45)
evalDeferred "shaderBallRendererMenuUpdate";
// Warning: line 1: filePathEditor: Attribute 'aiStandIn.dso' and label 'Standin' have been saved already. // 
// Warning: line 1: filePathEditor: Attribute 'aiVolume.filename' and label 'VDB' have been saved already. // 
// Warning: line 1: filePathEditor: Attribute 'aiImage.filename' and label 'Image' have been saved already. // 
// Warning: line 1: filePathEditor: Attribute 'aiPhotometricLight.aiFilename' and label 'IES' have been saved already. // 
import arnold
// Successfully imported python module 'arnold'
import mtoa
// Successfully imported python module 'mtoa'
import mtoa.cmds.registerArnoldRenderer;mtoa.cmds.registerArnoldRenderer.registerArnoldRenderer()
// Successfully registered renderer 'arnold'
updateRenderOverride;
// AbcImport v1.0 using Alembic 1.7.5 (built Feb  7 2018 15:59:45)
updateRendererUI;
// Breakdown for file: /home/sue.sauer/Downloads/barrelTests/Library/Assets/Barrels/BarrelsAsset.usd
[100%](1.59534e+09S) LoadStage
  [100%](1.59534e+09S) PostLoadProcess
    [100%](1.59534e+09S) PostImportLogic
    [100%](1.59534e+09S) CreatePrims
    [100%](1.59534e+09S) CreateTransformChains
    [100%](1.59534e+09S) HuntForNativePrims
  [100%](1.59534e+09S) OpeningUsdStage
    [100%](1.59534e+09S) ResetGlobalVariantFallbacks
    [100%](1.59534e+09S) UsdStageOpen
    [100%](1.59534e+09S) UpdateGlobalVariantFallbacks
    [100%](1.59534e+09S) OpenRootLayer
    [100%](1.59534e+09S) OpeningSessionLayer

// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithLid/repacked_BarrelWithLid_UV_<UDIM>_BaseColor.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithLid/repacked_BarrelWithLid_UV_<UDIM>_Metallic.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithLid/repacked_BarrelWithLid_UV_<UDIM>_Roughness.png // 
// Warning: [arnold-usd] Render was aborted. // 
select -r AL_usdmaya_Proxy ;
select -cl  ;
select -r AL_usdmaya_Proxy ;
file -f -new;
// untitled // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithLid/repacked_BarrelWithLid_UV_<UDIM>_BaseColor.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithLid/repacked_BarrelWithLid_UV_<UDIM>_Metallic.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithLid/repacked_BarrelWithLid_UV_<UDIM>_Roughness.png // 
select -cl  ;
// Warning: [arnold-usd] Render was aborted. // 
file -f -new;
// Clearing USD Stage Cache
// untitled // 
file -f -new;
// Clearing USD Stage Cache
// untitled // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithLid/repacked_BarrelWithLid_UV_<UDIM>_BaseColor.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithLid/repacked_BarrelWithLid_UV_<UDIM>_Metallic.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithLid/repacked_BarrelWithLid_UV_<UDIM>_Roughness.png // 
// Warning: [arnold-usd] Render was aborted. // 
select -cl  ;
// Warning: [arnold-usd] Render was aborted. // 
select -r BarrelsAsset_usd ;
select -cl  ;
file -f -new;
// Clearing USD Stage Cache
// untitled // 
// Warning: Selected hydra renderer doesn't support prim type 'DistantLight' // 
// Warning: Selected hydra renderer doesn't support prim type 'DiskLight' // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_TallJarWithHerbs3/repacked_TallJarWithHerbs3_<UDIM>_BaseColor.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_TallJarWithHerbs3/repacked_TallJarWithHerbs3_<UDIM>_Metallic.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_TallJarWithHerbs3/repacked_TallJarWithHerbs3_<UDIM>_Roughness.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_TallJarWithHerbs2/repacked_TallJarWithHerbs2_<UDIM>_BaseColor.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_TallJarWithHerbs2/repacked_TallJarWithHerbs2_<UDIM>_Metallic.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_TallJarWithHerbs2/repacked_TallJarWithHerbs2_<UDIM>_Roughness.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_repacked_BasketOfCarpets/repacked_BasketOfCarpets_UV_<UDIM>_BaseColor.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_repacked_BasketOfCarpets/repacked_BasketOfCarpets_UV_<UDIM>_Metallic.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_repacked_BasketOfCarpets/repacked_BasketOfCarpets_UV_<UDIM>_Roughness.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithNoLid/repacked_BarrelWithNoLid_UV_<UDIM>_BaseColor.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithNoLid/repacked_BarrelWithNoLid_UV_<UDIM>_Metallic.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithNoLid/repacked_BarrelWithNoLid_UV_<UDIM>_Roughness.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithNoLid_v2/BarrelWithNoLid_UV_<UDIM>_BaseColor.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithNoLid_v2/BarrelWithNoLid_UV_<UDIM>_Metallic.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithNoLid_v2/BarrelWithNoLid_UV_<UDIM>_Roughness.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithLid/repacked_BarrelWithLid_UV_<UDIM>_BaseColor.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithLid/repacked_BarrelWithLid_UV_<UDIM>_Metallic.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_BarrelWithLid/repacked_BarrelWithLid_UV_<UDIM>_Roughness.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_TallJarWithHerbs3/repacked_TallJarWithHerbs3_<UDIM>_BaseColor.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_TallJarWithHerbs3/repacked_TallJarWithHerbs3_<UDIM>_Metallic.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_TallJarWithHerbs3/repacked_TallJarWithHerbs3_<UDIM>_Roughness.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_TallJarWithHerbs2/repacked_TallJarWithHerbs2_<UDIM>_BaseColor.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_TallJarWithHerbs2/repacked_TallJarWithHerbs2_<UDIM>_Metallic.png // 
// Warning: Unable to find UDIM tiles for ../Textures/repacked_TallJarWithHerbs2/repacked_TallJarWithHerbs2_<UDIM>_Roughness.png // `

![Capture](https://user-images.githubusercontent.com/58251706/88067998-ead72c80-cb6f-11ea-8167-82ec47bf6731.PNG)

Im on pizar usd 20.05
and maya usd  7e140670c46c333f46c1c6c81d0acbe2948b6b33

**Comment by HdC-adsk** [comment]
@suesauer Looks like you are not using VP2RenderDelegate? Can you find the message in the output window?

```
#############################################################################
#  VP2_RENDER_DELEGATE_PROXY is overridden to 'true'.  Default is 'false'.  #
#############################################################################
```

@williamkrick Any other suspect?

**Comment by kxl-adsk** [comment]
@suesauer You will need to adopt USD 20.08 in order to test UDIM support.

**Comment by suesauer** [comment]
Last time I tried adopting 20.08 the engine.cpp didn't want to build. Has that been fixed?

**Comment by suesauer** [comment]
FYI
![Capture](https://user-images.githubusercontent.com/58251706/88143859-37fee100-cbf8-11ea-9af5-32b214f6e296.PNG)

**Comment by kxl-adsk** [comment]
@suesauer We haven't yet merged #617 so in order to adopt USD 20.08 you need to:
1. grab the tip of the `dev` branch
2. merge #617 locally

**Comment by kxl-adsk** [comment]
UDIM support has been added some time ago and last two GitHub releases came with USD 20.08. Closing this ticket.
